src/Controller/IslandoraWorkbenchIntegrationFileRenameController.php

```diff
@@ -0,0 +1,181 @@
+     */
+    protected $eventDispatcher;
+
+
+    /**
+     * Constructor to inject the Symfony Filesystem service.
+     */
+    public function __construct(FileSystemInterface $fileSystem, EventDispatcherInterface $event_dispatcher)
+    {
+        $this->fileSystem = $fileSystem;
+        $this->eventDispatcher = $event_dispatcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function create(ContainerInterface $container)
+    {
+        return new static(
+            $container->get('file_system'),
+            $container->get('event_dispatcher')
+        );
+    }
+
+
+
+    /**
+     * Renames a file while keeping the extension unchanged.
+     *
+     * Expects a URL parameter for the file ID (fid) and a JSON payload with:
+     * - new_filename: The new base file name (without extension).
+     *
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *   The incoming request.
+     * @param Drupal\file\Entity\File                   $file
+     *   The file ID from the URL.
+     *
+     * @return \Symfony\Component\HttpFoundation\JsonResponse
+     *   The JSON response.
+     */
+    public function main(Request $request, $file)
+    {
+        // Decode JSON payload.
+
+        $data = json_decode($request->getContent(), true);
+
+        // Validate that new_filename is provided.
+        if (empty($data['new_filename'])) {
+            return new JsonResponse(['error' => 'Missing new_filename parameter']);
+        }
+
+
+        if (!$file) {
+            return new JsonResponse(["Error 404" => 'No file found']);
+        }
+
+
+        $pathinfo = pathinfo($file->getFileUri());
+        $filename_new = $data['new_filename'];
+        $msg = $this->validate($file, $filename_new);
+        if ($msg != 'OK') {
+            return new JsonResponse(['Error' => $msg]);
+        }
+
+
+
+        if ($filename_new != $file->getFilename()) {
+            $filepath_new = $this->getRenamedFilePath($file, $filename_new);
+            rename($file->getFileUri(), $filepath_new);
+            // Update file entity.
+            $file->setFilename($filename_new);
+            $file->setFileUri($filepath_new);
+            $status = $file->save();
+
+            return new JsonResponse(['message' => 'File renamed successfully', 'new_filename' => $filename_new]);
+        }
+
+    }
+
+
+    /**
+     * Check if the file name is valid
+     *
+     * @param Drupal\file\Entity\File $file
+     *   File that needs to be renamed
+     *
+     * @param string                  $new_filename
+     *   File Name of the new file
+     * 
+     * @return string
+     *   Error Message if not valid else OK
+     */
+    public function validate($file, $new_filename)
+    {
+        $pathinfo = pathinfo($file->getFileUri());
+        $source_file_uri = $file->getFileUri();
+
+        if (!file_exists($source_file_uri)) {
+            // Show an error if no file on disc.
+            return 'No file exists in this';
```

Comment:
It would be useful to have this error message say "No file with name XX exists in the XXX filesystem."